add isNumeric to strutils

lihf8515 opened the pull request:
Add isNumeric()

lihf8515 committed:
Update strutils.nim

Add isNumeric()

ringabout commented:
Please add some testcases or `runnableExamples`.

lihf8515 committed:
Update strutils.nim

Add `isNumeric()` runnableExamples.

ringabout commented:
tests/gc/thavlak.nim is flasky in windows, it is third time for me to encounter this issue

```
2020-11-10T10:52:11.6660832Z FAIL: tests/gc/thavlak C  --gc:markAndSweep
2020-11-10T10:52:11.6706287Z Test "tests\gc\thavlak" in category "gc"
2020-11-10T10:52:11.6882905Z Failure: reOutputsDiffer
2020-11-10T10:52:11.6951124Z Expected:
2020-11-10T10:52:11.7113039Z Welcome to LoopTesterApp, Nim edition
2020-11-10T10:52:11.7154667Z Constructing Simple CFG...
2020-11-10T10:52:11.7191776Z 5000 dummy loops
2020-11-10T10:52:11.7246209Z Constructing CFG...
2020-11-10T10:52:11.7331905Z Performing Loop Recognition
2020-11-10T10:52:11.7480769Z 1 Iteration
2020-11-10T10:52:11.7481749Z Another 3 iterations...
2020-11-10T10:52:11.7539547Z ...
2020-11-10T10:52:11.7584938Z Found 1 loops (including artificial root node) (3)
2020-11-10T10:52:11.7621064Z 
2020-11-10T10:52:11.7654304Z Gotten:
2020-11-10T10:52:11.7776045Z Welcome to LoopTesterApp, Nim edition
2020-11-10T10:52:11.7917028Z Constructing Simple CFG...
2020-11-10T10:52:11.8265282Z 5000 dummy loops
2020-11-10T10:52:11.8543355Z Constructing CFG...
2020-11-10T10:52:11.9532232Z Performing Loop Recognition
2020-11-10T10:52:12.0216824Z 1 Iteration
2020-11-10T10:52:12.1075499Z Another 3 iterations...
2020-11-10T10:52:12.1693634Z ...
2020-11-10T10:52:12.2351723Z 
```

narimiran commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,59 @@ proc isDigit*(c: char): bool {.noSideEffect,
+  ## Check whether the string is numeric.
+  ## When the string is an integer, float or exponential, it returns true,
+  ## otherwise it returns false.
+  ## Note: The reason why `parseFloat()` is not used to achieve this is its
```
If this note should be part of public documentation, add one blank comment line before it, and bold it:

```suggestion
  ##
  ## **Note:** The reason why `parseFloat()` is not used to achieve this is its
```

xflywind replied:
```suggestion
  ## Note: The reason that `parseFloat()` is not used to achieve this is its
```

narimiran commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,59 @@ proc isDigit*(c: char): bool {.noSideEffect,
+  ## When the string is an integer, float or exponential, it returns true,
+  ## otherwise it returns false.
+  ## Note: The reason why `parseFloat()` is not used to achieve this is its
+  ## poor performance.
```
Also, can you back up your claim about poor performance of `parseFloat` with some numbers? Can we improve it?

ringabout commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,59 @@ proc isDigit*(c: char): bool {.noSideEffect,
   return c in Digits
 
+proc isNumeric*(s: string): bool =
+  ## Check whether the string is numeric.
```
```suggestion
  ## Checks whether the string is numeric.
```

lihf8515 committed:
Update lib/pure/strutils.nim

Co-authored-by: flywind <43030857+xflywind@users.noreply.github.com>

lihf8515 committed:
Update lib/pure/strutils.nim

Co-authored-by: Miran <narimiran@disroot.org>

lihf8515 commented:
Below is my source code：

``` nim
import times
import parseutils

proc isNumeric*(s: string): bool =
  var dotCount, plusCount, lessCount, eCount, numCount = 0
  for c in s:
    case c
    of '+':
      if eCount == 1:
        if plusCount == 2:
          return false
      else:
        if plusCount == 1:
          return false
      inc(plusCount)
    of '-':
      if eCount == 1:
        if lessCount == 2:
          return false
      else:
        if lessCount == 1:
          return false
      inc(lessCount)
    of '.':
      if dotCount == 1:
        return false
      inc(dotCount)
    of 'e', 'E':
      if numCount > 0 or dotCount > 0:
        if eCount == 1:
          return false
      else:
        return false
      inc(eCount)
    of '0'..'9':
      inc(numCount)
    else:
      return false

  return true

proc isNum*(s: string): bool =
  var rv = 0.0
  let L = parseFloat(s, rv, 0)
  if L != s.len or L == 0:
    return false
  return true

# Performance Testing
var t0 = cputime()
for i in 0..100000000:
  discard isNum("-10.56E+5")
var t1 = cputime() - t0
echo "isNum() time consuming:",t1

var t00 = cputime()
for i in 0..100000000:
  discard isNumeric("-10.56E+5")
var t11 = cputime() - t00
echo "isNumeric() time consuming:",t11
```

Use `d:release` to compile and run, the performance test is as follows:

Hint: C:\Users\Administrator\Desktop\f.exe  [Exec]
isNum() time consuming:4.271
isNumeric() time consuming:1.594

lihf8515 committed:
Update strutils.nim

lihf8515 committed:
Update strutils.nim

AmjadHD commented:
should `isNumeric(".9")` return `true` ?

juancarlospaco commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,54 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert isNumeric("e123.45") == false
+    doAssert isNumeric("123abc") == false
+    doAssert isNumeric("123.45.6") == false
+    doAssert isNumeric("123.45e++5") == false
```
```suggestion
    doAssert isNumeric("123.45e++5") == false
    doAssert isNumeric("Inf")
    doAssert isNumeric("-Inf")
    doAssert isNumeric("NaN")
```

- https://play.nim-lang.org/#ix=2DFU

I think `isNumeric` can be `func`, even with `--experimental:strictFuncs` compiles.

xflywind replied:
IMO `doAssert isNumeric(...) == false` is not idiom in Nim world. It should be `doAssert not isNumeric(...)`

AmjadHD commented:
```nim
func isNumeric2*(s: string): bool =
  var e, dot = false
  if s.len > 0:
    case s[0]
    of '-', '+':
      if s.len > 1:
        if s[1] notin {'0'..'9'}: return false
        if s.len == 2: return true
      else: return false
    of '0'..'9':
      if s.len > 1:
        if s.len == 2: return s[1] in {'0'..'9'}
        case s[1]
        of '0'..'9': discard
        of '.': dot = true
        of 'e', 'E': e = true
        else: return false
      else: return true
    else: return false
  else: return false
  
  var i = 2
  var c: char
  while i < s.len:
    c = s[i]
    case c
    of '0'..'9': discard
    of '.':
      if dot or e or i == s.high: return false
      dot = true
    of 'e', 'E':
      if e or i == s.high: return false
      e = true
      inc i
      case s[i]
      of '-', '+':
        if i == s.high or (i <= s.high and s[i+1] notin {'0'..'9'}): return false
      of {'0'..'9'}: discard
      else: return false
    else: return false

    inc i

  return true
```

```
isNumeric() time consuming:0.641
isNumeric2() time consuming:0.345
```
```
echo isNumeric("0.1")," = ",isNumeric2("0.1")  # true = true
echo isNumeric(".1")," = ",isNumeric2(".1")  # true = false
echo isNumeric("1.")," = ",isNumeric2("1.")  # true = false
```

lihf8515 commented:
@AmjadHD I tested your code and it is indeed faster than my code when `-d:danger` is not used. In addition, I think `.9` should be a numeric value and should not return `false`.

AmjadHD commented:
>it is indeed faster than my code when `-d:danger` is not used.

yeah, I didn't understand why is your version faster when `danger` is set
```
isNumeric() time consuming:0.0
isNumeric2() time consuming:0.22
```
>I think `.9` should be a numeric value and should not return `false`.

It depends on whether you're follwing nim's definition of a number. in Nim `.9` is not a valid number.

lihf8515 commented:
> 
> 
> > it is indeed faster than my code when `-d:danger` is not used.
> 
> yeah, I didn't understand why is your version faster when `danger` is set
> 
> ```
> isNumeric() time consuming:0.0
> isNumeric2() time consuming:0.22
> ```
> 
> > I think `.9` should be a numeric value and should not return `false`.
> 
> It depends on whether you're follwing nim's definition of a number. in Nim `.9` is not a valid number.

With the following code, it can be converted to 0.9, indicating that `.9` can be considered as a number in the form of a string.
```nim
var rv = 0.0
let L = parseFloat(".9", rv, 0)
echo rv
```

lihf8515 committed:
Update strutils.nim

ringabout commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,61 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert isNumeric("Inf") == true
+    doAssert isNumeric("-Inf") == true
+    doAssert isNumeric("NaN") == true
+  if s == "Inf" or s == "inf" or s == "-Inf" or s == "-inf" or s == "NaN" or s == "nan": return true
```
You should use `toLowerAscii` instead

```nim
case toLowerAscii(s)
of "inf", "nan", "-inf":
```
Or you should check it in while loop

lihf8515 replied:
I have considered using `toLowerAscii()`, but its performance loss is too great.

juancarlospaco replied:
https://play.nim-lang.org/#ix=2DJt

ringabout replied:
Your if statement is wrong.
NAn or nAN should be considered too.

ringabout replied:
you could check it in while loop for performance reason.

juancarlospaco replied:
Other way is to use something like `strict: static bool = false` then you can use `when strict:`...

timotheecour replied:
I don't think `static` makes a difference in practice in this particular case because of branch prediction; at least a test is needed if you think otherwise

lihf8515 committed:
Update strutils.nim

lihf8515 committed:
Update strutils.nim

Add isNumeric() to strutils.nim

lihf8515 committed:
Add isNumeric() to strutils.nim

ringabout commented:
Please consider applying following tests to `isNumeric`

should be true:
```nim
import strutils
doAssert "-.1".parseFloat == -0.1
doAssert "0.000_005".parseFloat == 5.000_000e-6
doAssert "1e1_00".parseFloat == "1e100".parseFloat

echo "0.00_0001".parseFloat(), " : ", 1E-6
echo "0.00__00_01".parseFloat(), " : ", 1E-6
echo "0.0_01".parseFloat(), " : ", 0.001
echo "0.00_000_1".parseFloat(), " : ", 1E-6
echo "0.00000_1".parseFloat(), " : ", 1E-6

echo "1_0.00_0001".parseFloat(), " : ", 10.000001
echo "1__00.00_0001".parseFloat(), " : ", 1_00.000001

```

should be false:
```nim
expect_fail("1_0._00_0001".parseFloat())
expect_fail("_1_0_00.0001".parseFloat())
expect_fail("10.00.01".parseFloat())
expect_fail("10.00E_01".parseFloat())
expect_fail("10.00E_01".parseFloat())
expect_fail("10.00E".parseFloat())
expect_fail("10.00A".parseFloat())
```

lihf8515 committed:
Update strutils.nim

Vindaar commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,107 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert not isNumeric("123.45.6") == true
+    doAssert not isNumeric("123.45e++5") == true
+    doAssert not isNumeric("5.2+e1") == true
+    doAssert isNumeric(".9") == true
```
Personally don't think it's a good idea to treat these numbers as numbers (by default).

lihf8515 replied:
I have added the parameter 'enabledNaN' to indicate whether special forms of values such as' NaN ', 'Inf', '.9 ', '9.' are valid.

Vindaar commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,107 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert isNumeric(".9") == true
+    doAssert isNumeric("Inf") == true
+    doAssert isNumeric("-Inf") == true
+    doAssert isNumeric("NaN") == true
```
Same for NaN. Here I think maybe an option should allow to choose whether NaN is allowed or not. If the user wishes to check whether something is a valid number, `isNumeric` would currently not be reliable. The same is true for Inf to a lesser degree.

Vindaar commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,107 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert isNumeric("-Inf") == true
+    doAssert isNumeric("NaN") == true
+    doAssert isNumeric("nAn") == true
+    doAssert isNumeric("-.1") == true
```
Same as `.9`.

lihf8515 committed:
Update strutils.nim

timotheecour commented:
```
Use nim c -r -d:release -d:danger test.nim to compile and run, the performance test is as follows:
isNum() time consuming:1.24
isNumeric() time consuming:8.890457814381136e-018
```
* I'm not saying isNumeric isn't faster, but this benchmark is flawed in several ways and those numbers should've given you an indication there was something wrong.
`discard isNumeric("5.2+e1")` can get optimized away; at least doAssert should be used, or, if you want even more precise measurement without incurring overhead from instruction cache, something like:

```nim
var a = 0
if isNumeric("5.2+e1"): a+=1
if isNumeric("5.2+e1"): a+=1
...
# and at the end:
doAssert a == execptedValue
```

* benchmark tests should be inside a proc, not at module scope (there is overhead in some cases)
* you may get artificial unrealistic optimizations from running in a loop those few hardcoded tests

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,122 @@ proc isDigit*(c: char): bool {.noSideEffect,
+  ## Checks whether the string is numeric.
+  ## When the string is an integer, float or exponential, it returns true,
+  ## otherwise it returns false.
+  ## 'enabledNaN' variable value indicates whether 'NaN', 'Inf', '.9 ', '9.'
```
`enabledNaN` is a bad name for that; maybe more granular?
enableNaNInf, enableLeadingDot, enableTrailingDot ?

ringabout replied:
I vote for `strict`
https://github.com/nim-lang/Nim/pull/15903#discussion_r521112176

timotheecour replied:
I'd rather have a minimum of flexibility than 1 option fits all (in practice this means 1-fits-none), so it can be used in more contexts eg:
* supports whatever json considers as numbers (see https://restfulapi.net/json-data-types/)
* supports whatever nim considers numbers
* accept vs support loose wrt preceding 0 (eg: `010`, which json rejects)
* accept vs support loose wrt `dot` (eg: `5.`, and `.5`)

and the simplest is either a `set[Options]` or a set of bool flags like `enableNaNInf, enableLooseDot, enableLooseZero` or other partitioning

lihf8515 committed:
Update strutils.nim

lihf8515 committed:
Update strutils.nim

lihf8515 committed:
Update strutils.nim

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
+  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
+  ## poor performance.
+  runnableExamples:
+    doAssert isNumeric("123") == true
```
doAssert isNumeric("123") == true
=>
doAssert isNumeric("123")

doAssert not isNumeric("e123.45") == true
=>
doAssert not isNumeric("e123.45")

(+everywhere relevant)

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -198,18 +200,18 @@ proc isNumeric*(s: string, enabledNaN = false): bool =
     if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
        (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
-      if enabledNaN:
+      if enableNaNInf:
```
return enableNaNInf
(+ elsewhere relevant)

juancarlospaco commented:
Nice amount of `runnableExamples`, I would suggest to put a few of those in a `static:` block, 
to make sure someone else can not break it by accident in the future.

timotheecour commented:
> Nice amount of runnableExamples
> I would suggest to put a few of those in a static: block

some might say excessive but we also want to convey semantics of what's accepted; in anycase anything that's obvious for readers from existing examples but needs testing should go to a dedicated test in tstrutils instead of runnableExamples; including `static: block` which can be done in tstrutils using the classical:

```nim
template main=
  doAssert ...
  doAssert ...

static: main()
main()
```

juancarlospaco commented:
But having a few on the `runnableExamples` makes it explicitly documented that it works at compile time, 
and most probably Nimscript and Javascript too, is one of the things I look for in `runnableExamples`.

timotheecour commented:
> But having a few on the runnableExamples makes it explicitly documented that it works at compile time, and most probably Nimscript and Javascript too, is one of the things I look for in runnableExamples.

ok but couldn't we instead improve `nim doc` so that it automatically tests which backends some proc works on, amongst: c,cpp,js,vm,nimscript (I'm not saying that's easy, but I'm saying that's easier , more useful, and scales better than having to rely on manual annotations in runnableExamples, which will never be kept up to date)

juancarlospaco commented:
@timotheecour 
I like the idea, `runnableExamples` already takes `-b:c` etc maybe it can take `-b:c,cpp,js` or  `-b:all`,
or maybe it can be a `template` inside `runnableExamples` that does the same as your code example,
and generate a comment on the HTML, something like:

:heavy_check_mark: C
:heavy_check_mark: C++
:negative_squared_cross_mark: JavaScript
:negative_squared_cross_mark: NimScript

AmjadHD commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
     doAssert isDigit('8') == true
   return c in Digits
 
+proc isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
```
Can't this be made a `func` ?
```suggestion
func isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
```

lihf8515 committed:
Update lib/pure/strutils.nim

Co-authored-by: Amjad Ben Hedhili <amjadhedhili@outlook.com>

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
+    doAssert not isNumeric("10.00E") == true
+    doAssert not isNumeric("10.00E+") == true
+    doAssert not isNumeric("10.00E-") == true
+    doAssert not isNumeric("10.00E_") == true
```
many of these belong instead in tstrutils.nim

lihf8515 committed:
Update strutils.nim

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
+
+  var i = 0
+  var eLeft, eRight, dot, e, num = false
+  while i < length:
```
`for i, si in s: ...` leads to clearer code

lihf8515 replied:
I've tested it multiple times on Windows and Linux, and the 'while' statement seems to be more efficient.

timotheecour replied:
that's troubling; do you have a reduced test case? because a lot of code relies on `for` being 0-cost compared to `while`

see also my comment above regarding unfair benchmark https://github.com/nim-lang/Nim/pull/15903#issuecomment-725520752

bung87 replied:
may use .mitems or .mpairs for getting less cost when us `for`.

lihf8515 replied:
Instead of a while statement, it's more efficient to use the form 'for i countup(0, sHigh)'.

juancarlospaco replied:
Nim compiles `for` to `while` anyways, so may not be the loop itself.

timotheecour replied:
this could be unrelated to while and instead related to:
```nim
for i, si in s:
  # use si
# vs
for i in low .. high:
  let si = s[i]
  # use si
```

but I don't understand the perf difference (not sure if related to pairs vs mpairs, etc)

lihf8515 replied:
I think it is associated with pairs vs MPairs.

timotheecour replied:
it can't be, s is a let variable.

I still don't trust the benchmark you did (see above remarks); if there is a difference, can you please show a valid benchmark showing the difference?

these give identical times:

nim r -d:danger -d:case2a --gc:arc main
2.296005964279175
nim r -d:danger -d:case2b --gc:arc main
same + noise

(ditto without --gc:arc)

```nim
when true:
  import times
  proc fn(s: string): int =
    var ret = 0
    when defined case2a:
      for i, si in s:
        ret += i
        ret += cast[int](si)
    when defined case2b:
      for i in 0..<s.len:
        ret += i
        ret += cast[int](s[i])
    return ret

  proc main =
    let ns = 1_000
    let n = 10_000_000
    var s: string
    for i in 0..<ns:
      s.add cast[char](i)
    var tot = 0
    for i in 0..<n:
      tot += fn(s)
    doAssert tot != -1

  let t = epochTime()
  main()
  echo epochTime() - t
```

note: modifying above code with `s: var string` and pairs vs mpairs would show a 2.5x slowdown, but as i said that's not what's at play here)

lihf8515 replied:
``` nim
import times

func isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
  ## Checks whether the string is numeric.
  ## When the string is an integer, float or exponential, it returns true,
  ## otherwise it returns false.
  ## The `enableNaNInf` value indicates whether the `NaN` and `Inf` values
  ## are valid.
  ## The `enableLooseDot` value indicates whether loose point values such as
  ## `.9` and `9.` are valid. 
  ##
  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
  ## poor performance.
  let length = s.len
  if length == 3:
    if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
       (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false
  if length == 4:
    if (s[0] in {'+', '-'} and s[1] in {'i', 'I'} and s[2] in {'n', 'N'} and s[3] in {'f', 'F'}) or
       (s[0] in {'+', '-'} and s[1] in {'n', 'N'} and s[2] in {'a', 'A'} and s[3] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false

  var eLeft, eRight, dot, e, num = false
  let sHigh = s.len - 1
  for i in countup(0, sHigh):
    case s[i]
    of '+', '-':
      if i == sHigh:
        return false
      if e == false:
        if num:
          return false
        if eLeft:
          return false
        eLeft = true
      else:
        if num:
          return false
        if eRight:
          return false
        eRight = true
    of '.':
      if dot:
        return false
      if not enableLooseDot:
        if num == false:
          return false
        else:
          if i == sHigh:
            return false
          if s[i+1] in {'e', 'E'}:
            return false
      num = false
      dot = true
    of 'e', 'E':
      if i == sHigh:
        return false
      if num or dot:
        if e:
          return false
        num = false
      else:
        return false
      e = true
    of '0'..'9':
      num = true
    of '_':
      if num == false:
        return false
      if dot and num == false:
        return false
      if e and num == false:
        return false
    else:
      return false

  return true

func isNumeric2*(s: string, enableNaNInf, enableLooseDot = false): bool =
  ## Checks whether the string is numeric.
  ## When the string is an integer, float or exponential, it returns true,
  ## otherwise it returns false.
  ## The `enableNaNInf` value indicates whether the `NaN` and `Inf` values
  ## are valid.
  ## The `enableLooseDot` value indicates whether loose point values such as
  ## `.9` and `9.` are valid. 
  ##
  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
  ## poor performance.
  let length = s.len
  if length == 3:
    if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
       (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false
  if length == 4:
    if (s[0] in {'+', '-'} and s[1] in {'i', 'I'} and s[2] in {'n', 'N'} and s[3] in {'f', 'F'}) or
       (s[0] in {'+', '-'} and s[1] in {'n', 'N'} and s[2] in {'a', 'A'} and s[3] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false

  var eLeft, eRight, dot, e, num = false
  let sHigh = s.len - 1
  for i, si in s:
    case si
    of '+', '-':
      if i == sHigh:
        return false
      if e == false:
        if num:
          return false
        if eLeft:
          return false
        eLeft = true
      else:
        if num:
          return false
        if eRight:
          return false
        eRight = true
    of '.':
      if dot:
        return false
      if not enableLooseDot:
        if num == false:
          return false
        else:
          if i == sHigh:
            return false
          if s[i+1] in {'e', 'E'}:
            return false
      num = false
      dot = true
    of 'e', 'E':
      if i == sHigh:
        return false
      if num or dot:
        if e:
          return false
        num = false
      else:
        return false
      e = true
    of '0'..'9':
      num = true
    of '_':
      if num == false:
        return false
      if dot and num == false:
        return false
      if e and num == false:
        return false
    else:
      return false

  return true

proc testIsNumeric() =
  var a = 0
  let execptedValue = 160000000
  let t0 = cputime()
  for i in 0..9999999:
    if isNumeric("123"): a += 1
    if isNumeric("123.45"): a += 1
    if isNumeric("+123.45"): a += 1
    if isNumeric("123.45e-2"): a += 1
    if isNumeric("+123.45E-2"): a += 1
    if isNumeric("-123.45e2"): a += 1
    if not isNumeric("e123.45"): a += 1
    if not isNumeric("abc"): a += 1
    if not isNumeric("123abc"): a += 1
    if not isNumeric("123.45.6"): a += 1
    if not isNumeric("123.45e++5"): a += 1
    if not isNumeric("5.2+e1"): a += 1
    if not isNumeric(".9"): a += 1
    if not isNumeric("Inf"): a += 1
    if not isNumeric("-Inf"): a += 1
    if not isNumeric("NaN"): a += 1
  let t1 = cputime() - t0
  if a == execptedValue:
    echo "OK! isNumeric() time consuming:", t1

proc testIsNumeric2() =
  var a = 0
  let execptedValue = 160000000
  let t0 = cputime()
  for i in 0..9999999:
    if isNumeric2("123"): a += 1
    if isNumeric2("123.45"): a += 1
    if isNumeric2("+123.45"): a += 1
    if isNumeric2("123.45e-2"): a += 1
    if isNumeric2("+123.45E-2"): a += 1
    if isNumeric2("-123.45e2"): a += 1
    if not isNumeric2("e123.45"): a += 1
    if not isNumeric2("abc"): a += 1
    if not isNumeric2("123abc"): a += 1
    if not isNumeric2("123.45.6"): a += 1
    if not isNumeric2("123.45e++5"): a += 1
    if not isNumeric2("5.2+e1"): a += 1
    if not isNumeric2(".9"): a += 1
    if not isNumeric2("Inf"): a += 1
    if not isNumeric2("-Inf"): a += 1
    if not isNumeric2("NaN"): a += 1
  let t1 = cputime() - t0
  if a == execptedValue:
    echo "OK! isNumeric2() time consuming:", t1

testIsNumeric()
testIsNumeric2()
```
The time difference is small, but there is still a difference.
OK! isNumeric() time consuming:1.82
OK! isNumeric2() time consuming:2.25

timotheecour replied:
-d:release
OK! isNumeric() time consuming:2.303698
OK! isNumeric2() time consuming:2.077815


-d:danger
OK! isNumeric() time consuming:1.999999999999832e-06
OK! isNumeric2() time consuming:1.999999999999398e-06

only -d:danger is relevant for benchmarkings, not -d:release

so there is no difference AFAIK, and the simplest form should be used (`or i, si in s:`)


## note
this actually highlights something interesting, that --gc:arc is really slow here:
-d:danger --gc:arc
OK! isNumeric() time consuming:0.011745
OK! isNumeric2() time consuming:0.011417

-d:danger
OK! isNumeric() time consuming:2.000000000000265e-06
OK! isNumeric2() time consuming:1.999999999999398e-06

(but that doesn't change above conclusion)
=> tracking this here: https://github.com/nim-lang/Nim/issues/15965

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
     doAssert isDigit('8') == true
   return c in Digits
 
+func isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
```
talking about efficiency, `s/string/openArray[char]/`
(avoids copies in some cases)

lihf8515 committed:
Update strutils.nim

lihf8515 committed:
Update strutils.nim

I've tested it multiple times on Windows and Linux, and the 'while' statement seems to be more efficient.

lihf8515 committed:
Update strutils.nim

simonkrauter commented:
Previous discussion: https://github.com/nim-lang/Nim/issues/10761

juancarlospaco commented:
Maybe can be better to use something like `options: set[NumericOptions] = {useNaN, useInf, useLooseDot}` a set[enum].
